### DISCUSSION TITLE
Nim insted GScript

### DISCUSSION THREAD

**Discussion by kodkuce** [godotengine/godot-proposals#4022] [Scripting]
Why make user specific lang when you can use something syntax similar like Nim that is fast and does not lock to only godot

**Comment by azur-wolve** [comment]
@kodkuce they won't gonna replace GDScript with Nim.

There are plans for GDScript; in some future an option will be added to pre-compile GDScript code to C-machine code through LLVM.
So eventually the speed issue will be solved.

They choose to make a custom scripting language for the engine rather than Nim to have direct control over it without depend on third parties, and for integration with the engine.

**Reply by kodkuce** [reply]
i kinda know its a uber long shot but at least i wanted to express my opinion. Hopefully GDExtensions will be a thing
+ over it without depend on third parties << this one is nice

**Reply by theraot** [reply]
@kodkuce Is GDExtension not a thing yet? If it isn't it will be a thing. Anyway, I believe there was a Godot-Nim binding over GDNative, you can search for that. I don't know how usable it is, but if it isn't, at least it is precedent. It does not have to be Nim instead of GDScript, it can be Nim in addition to GDScript via GDExtension/GDNative.

**Comment by fenix-hub** [comment]
GDScript is a pretty intuitive and easy-to-learn language.
Honestly I don't know what would be the benefit of replacing the engine specific language with a third party language, when everybody already have the possibility to write bindings for almost any language - and anyway, there are already "faster" languages which are officialy supported to be used: C and C++
You can find everything related to this topic here
[scripting languages](https://docs.godotengine.org/en/stable/getting_started/step_by_step/scripting_languages.html)
[gdnative and binginds](https://docs.godotengine.org/en/stable/tutorials/scripting/gdnative/what_is_gdnative.html)

**Comment by noidexe** [comment]
For reference, here are the reasons for creating GDscript (From https://docs.godotengine.org/en/stable/about/faq.html#doc-faq-what-is-gdscript)


====
What were the motivations behind creating GDScript?
---------------------------------------------------

In the early days, the engine used the `Lua <https://www.lua.org>`__ scripting
language. Lua can be fast thanks to LuaJIT, but creating bindings to an object-oriented
system (by using fallbacks) was complex and slow and took an enormous
amount of code. After some experiments with `Python <https://www.python.org>`__,
it also proved difficult to embed.

The main reasons for creating a custom scripting language for Godot were:

1. Poor threading support in most script VMs, and Godot uses threads
   (Lua, Python, Squirrel, JavaScript, ActionScript, etc.).
2. Poor class-extending support in most script VMs, and adapting to
   the way Godot works is highly inefficient (Lua, Python, JavaScript).
3. Many existing languages have horrible interfaces for binding to C++, resulting in a large amount of
   code, bugs, bottlenecks, and general inefficiency (Lua, Python,
   Squirrel, JavaScript, etc.) We wanted to focus on a great engine, not a great number of integrations.
4. No native vector types (vector3, matrix4, etc.), resulting in highly
   reduced performance when using custom types (Lua, Python, Squirrel,
   JavaScript, ActionScript, etc.).
5. Garbage collector results in stalls or unnecessarily large memory
   usage (Lua, Python, JavaScript, ActionScript, etc.).
6. Difficulty integrating with the code editor for providing code
   completion, live editing, etc. (all of them). This is well-supported
   by GDScript.

GDScript was designed to curtail the issues above, and more.


====

**Comment by ghost** [comment]
I love gdscript -- I'd use it as a general language if I could -- but a lot of the stated reasons for keeping it don't make much sense given how miserably slow it is (1, 4, and 5 above, for example). The current 4.x alpha only increases speed marginally, and I doubt it's feasible to make it much better. C# is likely to be the choice of most serious developers in the future, despite the bugs.

See https://github.com/godotengine/godot/issues/36060 for arguments against llvm, etc.

On the other hand, I'd never heard of nim until yesterday. I don't see much to recommend it over other scripting languages, but maybe I'm missing something.

**Reply by atirutw** [reply]
There has been a [proposal for compiling GDScript to WebAssembly](https://github.com/godotengine/godot-proposals/issues/3370) for faster performance, but it's not planned for Godot 4(.x?)

**Reply by unfa** [reply]
I am curious about where exactly do you experience performance issues with GDScript?
I have never had issues with GDScript being too slow myself.

**Reply by ghost** [reply]
Well, for myself, the first project I tried to port to godot from love2d nearly failed because I was used to lua's speed. I was trying to calculate paths for lots of mobs -- lua could do several hundred easily, gdscript could manage a few dozen. Of course, I got around it by prioritizing the most important mobs and cutting back on the number of calculations, but I can't really say it was the same game afterward.

gdscript is fine for a lot of applications, and there are options for optimizing trouble spots, but a fledgling programmer might just get frustrated and look for easier alternatives. Personally, I'd rather stick with gdscript, but I can definitely see why others might not.

**Reply by unfa** [reply]
Thanks, that's telling me a lot. I haven't been doing any AI work yet.
BTW - are you processing pathfinding for all mobs every single frame?

**Reply by ghost** [reply]
In lua, I was doing it all after every player turn. In godot I had to queue the mobs up to process when activated by a timer. It can take several seconds to complete them all, but that's usually not an issue since it's a slower, turn-based game, and the mobs closer to the player get priority. It only gets behind when the player is moving fast.

In lua, the mobs would constantly attack each other and respawn when killed. In godot, I have to limit that to a radius around the player, which results in peculiar behavior near the border.

It's an unusual situation. None of my other projects ran into serious issues, but you know that every new programmer brings some concept that she wants to achieve, and if she hits a road-block, she's apt to look for another solution.

I don't miss much about love2d, but I do miss lua, despite its many issues.

**Reply by theraot** [reply]
If you are talking about path finding, are you using Godot's AStar or Godot's Navigation System? I would expect them to take care of the heavy lifting.

**Reply by ghost** [reply]
I haven't tried the recent backported pathfinding, and it might do a better job. My first attempt was several years ago -- I never finished it, and I intend to get back to it soon. The older (pathfinding) system didn't make a significant difference in my case, probably because everything (in my game) is confined to a square grid system, so doing the math is relatively easy.

Godot does build in a lot of functions that make scripting speed less of a liability. I like that.

Another issue I had (though minor), was doing a flood-fill of the map to check for unreachable areas. It took less than a second on a pc, but caused a noticeable delay when exported to the web. I could fill in unreachable areas after the game starts, assuming of course that I don't accidentally start the player in one.

**Reply by Zireael07** [reply]
@duane-r: I had a web game that used floodfill... and never had any issues. Unless your map is very large?

**Reply by setanarut** [reply]
this thread is about nim support not general chatting. please

**Comment by shayneoneill** [comment]
Nims a great little language, its a moon-language, but a good moon-language. As explained above however, GD-Script exists for a reason. And GD-Script is also a good moon language.  Its simple to learn, and quite comfortable for a pythonista (Compared to "Boo", the Python-like language that used to be Unity, which was often very uncomfortable for python users because while it shared syntax it just didnt think like python leading to a lot of confusion. Boo died for a reason. GD-Script kinda does in many respect think like python, and theres little about the language that would introduce unwanted surprised to someone coming from python.  

So I dont think its reasonable to expect GD-Script to be replaced. The language is part of Godots DNA, and its fine. 

Nim CAN be introduced as an extension and in fact thats exactly what people have done. No idea how well it works, but if it does, its an everyone-wins solution. https://github.com/pragmagic/godot-nim

**Comment by unfa** [comment]
I guess a solution here would be to use another language to take care of
the heavy lifting? I think that's what C# support is for.

**Reply by ghost** [reply]
Of course, a cynic might say, "Why not just start in C# and save a step?" But, I think we're gettting off-topic. :)

**Comment by noidexe** [comment]
Try using while instead of for and if..elif instead of match in tight
loops. Or unroll loops altogether if it's only a few cases.
It can make a noticeable difference.
Also if you can have the same behavior with a node or buch of nodes do
that. A call to a native function will always be faster than the same
algorithm in gdscript.
